### PR TITLE
Fix #1330: Solve ABA issue in the nrt_pop_meminfo_list.

### LINKED ISSUE #1330: SegFault using threads, nogil=True and many (re-) allocations
System: x86_64 GNU/Linux, LLVM 3.6.1, Python 3.4.3, llvmlite latest git, Numba latest git
(also reproducible with llvmlite 0.6.0, Numba 0.20.0)

Given the script:
```python
from sys import argv
from numpy import zeros
from numba import jit
from concurrent.futures import ThreadPoolExecutor
from multiprocessing import cpu_count


@jit(nopython=True, nogil=True)
def loop(loops, size):
    for _ in range(loops):
        zeros(size).sum()


if __name__ == "__main__":
    size = 1000 if len(argv) <= 1 else int(argv[1])
    loops = 1000 if len(argv) <= 2 else int(argv[2])
    jobs = 1000 if len(argv) <= 3 else int(argv[3])
    with ThreadPoolExecutor(cpu_count()) as executor:
        for _ in range(jobs):
            executor.submit(loop, loops, size)
```
I receive a SegFault with the following `gdb` output (on a machine with 24 HT cores) :
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffed5c7700 (LWP 27511)]
[New Thread 0x7fffecdc6700 (LWP 27512)]
[New Thread 0x7fffe7fff700 (LWP 27513)]
[New Thread 0x7fffe77fe700 (LWP 27514)]
[New Thread 0x7fffe6ffd700 (LWP 27515)]
[New Thread 0x7fffe67fc700 (LWP 27516)]
[New Thread 0x7fffe5ffb700 (LWP 27517)]
[New Thread 0x7fffe57fa700 (LWP 27518)]
[New Thread 0x7fffe4ff9700 (LWP 27519)]
[New Thread 0x7fffc3fff700 (LWP 27520)]
[New Thread 0x7fffc37fe700 (LWP 27521)]
[New Thread 0x7fffc2ffd700 (LWP 27522)]
[New Thread 0x7fffc27fc700 (LWP 27523)]
[New Thread 0x7fffc1ffb700 (LWP 27524)]
[New Thread 0x7fffc17fa700 (LWP 27525)]
[New Thread 0x7fffc0ff9700 (LWP 27526)]
[New Thread 0x7fffa3fff700 (LWP 27527)]
[New Thread 0x7fffa37fe700 (LWP 27528)]
[New Thread 0x7fffa2ffd700 (LWP 27529)]
[New Thread 0x7fffa27fc700 (LWP 27530)]
[New Thread 0x7fffa1ffb700 (LWP 27531)]
[New Thread 0x7fffa17fa700 (LWP 27532)]
[New Thread 0x7fffa0ff9700 (LWP 27533)]
[New Thread 0x7fff83fff700 (LWP 27534)]

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffa27fc700 (LWP 27530)]
nrt_pop_meminfo_list (list=0x7fffeeb0d7c0 <TheMSys>) at numba/runtime/nrt.c:63
```

If you have trouble reproducing the error, try changing the number of jobs, loops, threads or the array size. Whether and how fast I get the segmentation fault depends on these values.
The script runs fine when not using `nogil=True` or using processes instead of threads via `ProcessPoolExecutor`.

### PR DESCRIPTION
**Pull request by sklam**
Fixes #1330.
Revert back to lock base implementation but a smarter one.
Allow concurrent pop, concurrent push, but not concurrent push & pop.

### CODE REVIEW

File: numba/runtime/nrt.c [hunk 1]
```diff
@@ -52,8 +64,77 @@ typedef struct {
...
+        }
+    }
+}
+
+static
+MemInfo *nrt_pop_meminfo_list(MemInfoList *listst) {
     MemInfo *old, *repl, *head;
+    MemInfo * volatile * volatile list = &listst->list;
+
+    list_lock(listst, LOCK_POP);
```
Comment (pitrou) [inline numba/runtime/nrt.c]: This looks complicated and difficult to understand. Why can't you lock the list in an exclusive way and then do stuff on it unconditionally? It looks like you are both locking the list *and* still allowing concurrent modifications to it. This seems hard to reason about.

**Comment by pitrou** [comment]
By the way, if we are actually using locking, it may be better to bite the bullet and use system APIs (pthread mutexes under POSIX, critical sections under Windows).

**Comment by pitrou** [comment]
Now that I think of it, yet another option is to remove the freelists and to allocate the data inline with the meminfo structure.

**Comment by sklam** [comment]
We don't always allocate the data.  MemInfo may reference borrowed pointer from numpy.

**Comment by pitrou** [comment]
Yes, but the only case where allocation performance may matter is when we allocate the data ourselves...

**Comment by sklam** [comment]
That's true.  We can then avoid the freelist and just rely on malloc() and free().

**Pull request closed without merging**